src/hooks/transactions/useSignTransactions.tsx

```diff
@@ -1,189 +1,236 @@
-        }
+
+      const isProviderInitialized = await provider.init();
+
+      if (!isProviderInitialized) {
+        return;
       }
-    } catch (err) {
-      const errorMessage =
-        (error as unknown as Error)?.message ||
-        (error as string) ||
-        'error signing transaction';
-      console.error('error signing transaction', errorMessage);
+
+    } catch (error) {
+
+      const customErrorMessage = 'provider not intialized';
+      const errorMessage = (error as unknown as Error)?.message || (error as string) || customErrorMessage;
+      console.error(customErrorMessage, errorMessage);
       onCancel(errorMessage);
+      return;
+
+    };
+
+    try {
+
+      const signedTransactions = await provider.signTransactions(transactions);
+      const hasSameTransactions = Object.keys(signedTransactions).length === transactions.length;
+      const hasAllTransactionsSigned = signedTransactions && hasSameTransactions;
+      const shouldMoveTransactionsToSignedState = signedTransactions && hasAllTransactionsSigned;
+
+      if (!shouldMoveTransactionsToSignedState) {
+        return;
+      }
+
+      const signedTransactionsArray = Object.values(signedTransactions).map((tx) => parseTransactionAfterSigning(tx) );
+
+      dispatch(
+        moveTransactionsToSignedState({
+          sessionId,
+          transactions: signedTransactionsArray,
+          status: TransactionBatchStatusesEnum.signed
+        })
+      );
+
+      if (shouldRedirectAfterSign) {
+        window.location.href = redirectRoute;
+      }
+
+    } catch (err) {
+
+      const customErrorMessage = 'error signing transaction';
+      const errorMessage = (error as unknown as Error)?.message || (error as string) || customErrorMessage;
+      console.error(customErrorMessage, errorMessage);
+      onCancel(errorMessage, sessionId);
+    
+    };
+    
+  };
+
+  const signTransactions = async () => {
+
+    const hasSessionId = Boolean(transactionsToSign?.sessionId);
+    const { sessionId, transactions, callbackRoute } = transactionsToSign!;
+    const hasTransactions = Boolean(transactionsToSign?.transactions?.length);
+    const isSignEligible = transactionsToSign && hasSessionId && hasTransactions;
+    //move to errors constants
+    const missingProviderErrorMessage = 'You need a signer/valid signer to send a transaction,use either WalletProvider, LedgerProvider or WalletConnect';
+
+    if (!isSignEligible) {
+      return;
     }
-  }
 
-  useEffect(() => {
-    if (transactionsToSign?.sessionId) {
-      signTransactions();
+    if (!provider) {
+      console.error(missingProviderErrorMessage);
+      return;
+    }
+
+   
+    /*
+    * if the transaction is cancelled 
+    * the callback will go to undefined,
+    * we save the most recent one for a valid transaction
+    */
+    savedCallback.current = callbackRoute || window.location.pathname;
+
+    const signWithWallet = ( transactions: Array<Transaction> ) => {
```

Comment:
nonono, don't declare functions inside functions, this can create a bunch of problems and a strain on memory. I think a better way would be to pull them out and pass the necessary arguments